EIPS/eip-9999.md

```diff
@@ -0,0 +1,92 @@
+---
+eip: 9999
+title: Block-Level Access Lists
+description: Enforced block-level access lists for parallalizing transaction validation
+author: Toni Wahrstätter (@nerolation), Dankrad Feist (@dankrad), Francesco D`Amato (@fradamt), Yoav Weiss (@yoavw), Alex Forshtat (@forshtat)
+discussions-to: https://ethereum-magicians.org/t/eip-xxxx-block-level-access-lists/23337
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-03-31
+---
+
+## Abstract
+
+This EIP proposes adding block-level access lists. By including a complete list of all addresses and storage keys accessed during a block, along with their initial values, in the block header, we can improve execution efficiency and enable better parallelization.
+
+## Motivation
+
+Currently, transactions without an explicit transaction access list cannot be efficiently parallelized, as the execution engine cannot determine in advance which addresses and storage slots will be accessed. While transaction-level access lists exist, they are not enforced, making it difficult to optimize execution pipelines.
```

Comment:
The scope of this EIP seems to be parallelization by usage of access lists. The focus seems to be storage slots. Note that other things can also change, namely: account code and account balance. This is definitely needed for parallelization, think of an easy example where Tx 1 edits the balance of account A, and Tx 2 does (code) branch X if A has balance < C or branch Y if A has balance >= C. For this parallel execution balances would also need to be tracked.

Reply:
Yeah agree, this is needed. We could have addresses mapped to an ordered list of events which are tuples containing the tx index and another tuple with the pre and post balance:

```python
class BlockBody:
    ...
    # Mapping each address to a list of balance diff events.
    # Each event is a tuple containing:
    #   - The transaction index that effected the change.
    #   - The balance diff
    balance_diffs: Dict[
        Address,
        List[Tuple[int, Value]]
    ]
    ```